core/lib/Thelia/Controller/Admin/ProductController.php

```diff
@@ -1804,4 +1816,365 @@ protected function arrayHasEntries(array $data, array $entries)
 
         return $status;
     }
+
+    /*
+     * PRODUCT CLONING PROCESS
+     *
+     *  Vars prefixes :
+     * op = Original Product
+     * cp = Clone Product
+     */
+
+    /**
+     * @return mixed|\Symfony\Component\HttpFoundation\Response
+     * @throws \Exception
+     */
+    public function cloneAction()
+    {
+        if (null !== $response = $this->checkAuth($this->resourceCode, $this->getModuleCode(), [AccessManager::CREATE, AccessManager::UPDATE])) {
+            return $response;
+        }
+
+        // Initialize vars
+        $cpf = $this->createForm(AdminForm::PRODUCT_CLONE);
+        $lang = $this->getSession()->getLang()->getLocale();
+
+        try {
+            // Check the form against constraints violations
+            $form = $this->validateForm($cpf, "POST");
+
+            // Get the form field values
+            $data = $form->getData();
+
+            $opId = $data['productId'];
+
```

Comment:
Instead of coding all the logic in the controller, It would be better to put it in Thelia\Action\Product class, and send a PRODUCT_CLONE event, that will be processed by this class.

Furthermore, il will allow any module interested by product cloning to subscribe to this event, and perform some specific actions.